Report Connect usage events

gzdunek opened the pull request:
Part of https://github.com/gravitational/teleport/issues/19000 
Requires https://github.com/gravitational/teleport/pull/19378 to be merged first. ✅
Requires https://github.com/gravitational/teleport/pull/19774 to be merged first (see https://github.com/gravitational/teleport/pull/19564#issuecomment-1369049255) ✅

This PR adds actual events sending mechanism, it receives events from Electron, ~converts them to prehog model~ anonyzmies them and then sends in batches.

EDIT:
In the first attempt I had separate definitions for prehog events in tshd. It was ineffective approach and resulted in quite a lot of boilerplate (maintaining two, very similar protos and then having to convert one model to another). After a discussion with Rafał, we decided to use prehog protos directly in tshd protos.

gzdunek committed:
Make `UsageReporter` reusable, move it to the new package

gzdunek committed:
Add `TeleportUsageReporter`

gzdunek committed:
Rename `UsageSubmitFunc` and `SubmitAnonymizedUsageEvents`

gzdunek committed:
Rename `NewUsageReporterTeleport` to `NewTeleportUsageReporter`

gzdunek committed:
Report Connect usage events

gzdunek commented:
My thoughts about this PR:
Anonymizing and converting tshd events to prehog events introduces quite a lot of boilerplate (the events have almost the same properties and I need manually convert each of them). I’m considering getting rid of tshd's `usage_events.proto` and instead operating only prehog's events. 
It would require:
- Performing anonymization before event leaves electron - shouldn’t be difficult, I can add a gRPC call that will run our anonymizer against the provided value
- Importing prehog's `connect.proto` to teleterm's `service.proto` - for some reason it doesn’t work, I get an error saying that imported proto doesn't exist 
- Generating JS files for `connect.proto`

What do you think about it? Do you like the current solution, where we send `authClusterId` with every event that needs to be anonymized? I'm afraid it can be confusing, as we create almost complete events at Electron level :/

avatus commented on lib/teleterm/api/proto/v1/usage_events.proto:
```diff
@@ -0,0 +1,84 @@
+  ConnectUsageEventOneOf event = 3;
+}
+
+message EventReportedResponse {}
```
This can also be an `EmptyResponse`

gzdunek replied:
Done, I think it's ok to use our common `EmptyResponse` instead of `google.protobuf.Empty`.

avatus reviewed (commented):
I only had a chance to look at the types today. Will finish my review tomorrow. Super neat so far tho!

ravicious commented on lib/prehog/proto/prehog/v1alpha/connect.proto:
```diff
@@ -0,0 +1,98 @@
+  }
+}
+
+message SubmitConnectEventResponse {}
```
We could use [`google.protobuf.Empty`](https://cloud.google.com/apis/design/design_patterns#empty_responses) here, as long as it doesn't return linter errors (I'm not sure about this).

Historically, I haven't been consistent with it inside lib/teleterm. [tshd_events_service.proto](https://github.com/gravitational/teleport/blob/71b19fa61d504a7f4ea9fbe1536d622babb5b44e/lib/teleterm/api/proto/v1/tshd_events_service.proto) explicitly defines custom empty messages in order to keep backwards compatibility.

But I did a quick search right now and it turns out [reusing google.protobuf.Empty (or something similar) is not a problem for backwards compatibility](https://stackoverflow.com/questions/50993815/is-google-protobuf-empty-dangerous-for-backwards-compatibility). Eric Anderson from the gRPC team says that old clients will not have a problem with this and the only thing it breaks is the generated types.

So I think going forward we could use google.protobuf.Empty, as we already do in the project. Unless the linter is now set up to require you to define a custom message for each RPC.

espadolini replied:
The `.proto` is imported from [gravitational/prehog](https://github.com/gravitational/prehog), which enforces the default Buf lint rules [RPC_REQUEST_STANDARD_NAME and RPC_RESPONSE_STANDARD_NAME](https://docs.buf.build/lint/rules#rpc_request_standard_name-rpc_response_standard_name-rpc_request_response_unique).

ravicious commented on lib/teleterm/api/proto/v1/usage_events.proto:
```diff
@@ -0,0 +1,84 @@
+
+option go_package = "github.com/gravitational/teleport/lib/teleterm/v1";
+
+message ReportEventRequest {
```
As for importing connect.proto to service.proto:

> The protocol compiler searches for imported files in a set of directories specified on the protocol compiler command line using the -I/--proto_path flag. If no flag was given, it looks in the directory in which the compiler was invoked. In general you should set the --proto_path flag to the root of your project and use fully qualified names for all imports.

https://developers.google.com/protocol-buffers/docs/proto3#importing_definitions

AFAIK we now use `buf` instead of `protoc`, so maybe enabling that requires changes to buf's config.

gzdunek replied:
>Could the Electron app send prehog's events to tshd through TerminalService and then tshd would anonymize the required fields before passing the events further?

Unfortunately, it is not possible as tshd events (the ones that need to be anonymized) have one additional field - `auth_cluster_id`. It is needed to create the anonymizer. The prehog events don't have this field.

gzdunek replied:
I just pushed the changes, so we don't have tshd usage events anymore and there is much less boilerplate. 

We should merge changes to the protobuf generation first, @ravicious promised to create a PR for that :)

gzdunek committed:
Merge branch 'master' into gzdunek/make-usage-reporter-reusable

gzdunek committed:
Remove timestamp from test

gzdunek committed:
Update e-ref

gzdunek committed:
Fix imports

gzdunek committed:
Rename receivers

gzdunek committed:
Add comment

gzdunek committed:
Merge branch 'master' into gzdunek/make-usage-reporter-reusable

# Conflicts:
#	e

gzdunek committed:
Bring back original `MaxBatchAge` in tests

gzdunek committed:
Fix comment for retryAttempts

gzdunek committed:
Remove tshd usage events, opearate only on prehog protos

gzdunek commented on lib/teleterm/daemon/usagereporter.go:
```diff
@@ -0,0 +1,252 @@
+
+	// submitDelay is a mandatory delay added to each batch submission
+	// to avoid spamming the prehog instance.
+	retryAttempts = 1
```
Fixed (also the number of retries).

gzdunek commented on lib/teleterm/daemon/usagereporter.go:
```diff
@@ -0,0 +1,252 @@
+
+		// Note: the backend doesn't support batching at the moment.
+		for _, event := range events {
+			// Note: this results in retrying the entire batch, which probably
+			// isn't ideal.
+			req := connect.NewRequest(event.Event)
+			if _, err := client.SubmitConnectEvent(ctx, req); err != nil {
+				errors = append(errors, err)
+			}
```
No, it works this way for Teleport events too, we can fix it later.

ravicious commented:
I opened #19774 which reorganizes protos into a single workspace.

I made a couple of small changes compared to #19700. Once #19774 is merged, you will need to do the following in this PR:

- [ ] Add `lib/prehog/buf-teleterm.gen.yaml`

<details>
<summary>lib/prehog/buf-teleterm.gen.yaml</summary>

```yaml
# buf-teleterm.gen.yaml is identical to buf.gen.yaml,
# with the exception of three addidiotal JS plugins.
version: v1
managed:
  enabled: true
  go_package_prefix:
    default: github.com/gravitational/teleport/lib/prehog/gen
plugins:
  - name: go
    path: lib/prehog/bin/protoc-gen-go
    out: lib/prehog/gen
    opt:
      - paths=source_relative
  - name: connect-go
    path: lib/prehog/bin/protoc-gen-connect-go
    out: lib/prehog/gen
    opt:
      - paths=source_relative

  - name: js
    out: lib/prehog/gen-js
    opt:
      - import_style=commonjs,binary
  - name: grpc
    out: lib/prehog/gen-js
    opt: grpc_js
    path: grpc_tools_node_protoc_plugin
  - name: ts
    out: lib/prehog/gen-js
    opt: "service=grpc-node"
```
</details>

- [ ] Extend `grpc-teleterm/host` to generate prehog protos as well.

```make
grpc-teleterm/host: protos/all
	$(BUF) generate --template=lib/prehog/buf-teleterm.gen.yaml lib/prehog/proto
	$(BUF) generate --template=lib/teleterm/buf.gen.yaml lib/teleterm/api/proto
```

- [ ] Add `go_package` in `lib/prehog/proto/prehog/v1alpha/teleport.proto`

```proto
option go_package = "github.com/gravitational/teleport/lib/prehog/gen/prehog/v1alpha";
```

gzdunek committed:
Change `prehog_event` to `prehog_req`

gzdunek committed:
Merge branch 'master' into gzdunek/add-usage-reporter-to-connect

# Conflicts:
#	e

gzdunek committed:
Populate failed events array

gzdunek committed:
Rename things a little

gzdunek committed:
Fix lint issues

gzdunek committed:
Rename `ReportEvent` to `ReportUsageEvent` in gRPC service for better context

gzdunek committed:
Do not check license in generated prehog js files

gzdunek committed:
Update protos

gzdunek commented:
License check was failing on the following files:
```
lib/prehog/gen-js/prehog/v1alpha/teleport_grpc_pb.d.ts
lib/prehog/gen-js/prehog/v1alpha/connect_grpc_pb.d.ts
lib/prehog/gen-js/prehog/v1alpha/connect_pb.d.ts
lib/prehog/gen-js/prehog/v1alpha/teleport_pb.d.ts
```
I added `lib/prehog/gen-js/**`  to the ignored files list.

gzdunek committed:
Merge branch 'master' into gzdunek/add-usage-reporter-to-connect

gzdunek committed:
Merge branch 'master' into gzdunek/add-usage-reporter-to-connect

gzdunek committed:
Merge branch 'master' into gzdunek/add-usage-reporter-to-connect

gzdunek committed:
Update Connect proto files to match prehog

gzdunek committed:
Generate JS files for prehog protos

gzdunek committed:
Pass prehog address as argument to tshd process instead of hard-coding it

gzdunek commented:
I updated the PR: 
- updated `connect.proto` to match the final version from `prehog`
- added changes related to JS files generation (from https://github.com/gravitational/teleport/pull/19564#issuecomment-1369049255) 
- removed hardcoded prehog address, instead it will be passed from Electron (similarly to other runtime values)

ravicious commented on lib/prehog/proto/prehog/v1alpha/connect.proto:
```diff
@@ -0,0 +1,103 @@
+  string cluster_name = 1;
+  // anonymized
+  string user_name = 2;
+  bool is_upload = 3;
```
What's the advantage of doing it this way vs using a string with `"upload"/"download"`?

gzdunek replied:
I followed Edoardo suggestion, I think `isUpload` is easier to understand than `direction`.

ravicious commented on tool/tsh/tsh.go:
```diff
@@ -181,6 +181,8 @@ type CLIConf struct {
 	DaemonCertsDir string
 	// DatabaseService specifies the database proxy server to log into.
 	DatabaseService string
+	// DaemonPrehogAddr is the URL where prehog events should be submitted.
+	DaemonPrehogAddr string
```
Let's put it below `DaemonCertsDir` and above `DatabaseService`.

gzdunek replied:
Right, I meant to put it like that.

gzdunek committed:
Keep `DaemonPrehogAddr` with other prehog values

gzdunek committed:
Merge branch 'master' into gzdunek/add-usage-reporter-to-connect

gzdunek committed:
Merge branch 'master' into gzdunek/add-usage-reporter-to-connect

gzdunek committed:
Merge branch 'master' into gzdunek/add-usage-reporter-to-connect
